syntax = "proto3";
package semantic_engine;

service SemanticEngine {
    // Ingests a batch of triples
    rpc IngestTriples (IngestRequest) returns (IngestResponse);
    
    // Queries the graph (Basic traversal for now)
    rpc GetNeighbors (NodeRequest) returns (NeighborResponse);
    
    // Vector Search (Placeholder for hybrid query)
    rpc Search (SearchRequest) returns (SearchResponse);

    // Resolves a string URI to a Node ID
    rpc ResolveId (ResolveRequest) returns (ResolveResponse);
    
    // Get all stored triples (for graph visualization)
    rpc GetAllTriples (EmptyRequest) returns (TriplesResponse);

    // Executes a SPARQL query
    rpc QuerySparql (SparqlRequest) returns (SparqlResponse);

    // Deletes all data associated with a namespace
    rpc DeleteNamespaceData (EmptyRequest) returns (DeleteResponse);
}

message SparqlRequest {
    string query = 1;
    string namespace = 2;
}

message SparqlResponse {
    string results_json = 1;
}

message DeleteResponse {
    bool success = 1;
    string message = 2;
}

message Provenance {
    string source = 1;
    string timestamp = 2;
    string method = 3;
}

message Triple {
    string subject = 1;
    string predicate = 2;
    string object = 3;
    Provenance provenance = 4;
}

message IngestRequest {
    repeated Triple triples = 1;
    string namespace = 2;
}

message IngestResponse {
    uint32 nodes_added = 1;
    uint32 edges_added = 2;
}

message NodeRequest {
    uint32 node_id = 1;
    string namespace = 2;
}

message NeighborResponse {
    repeated Neighbor neighbors = 1;
}

message Neighbor {
    uint32 node_id = 1;
    string edge_type = 2;
}

message SearchRequest {
    string query = 1;
    uint32 limit = 2;
    string namespace = 3;
}

message SearchResponse {
    repeated SearchResult results = 1;
}

message SearchResult {
    uint32 node_id = 1;
    float score = 2;
    string content = 3;
}

message ResolveRequest {
    string content = 1;
    string namespace = 2;
}

message ResolveResponse {
    uint32 node_id = 1;
    bool found = 2;
}

message EmptyRequest {
    string namespace = 1;
}

message TriplesResponse {
    repeated Triple triples = 1;
}
